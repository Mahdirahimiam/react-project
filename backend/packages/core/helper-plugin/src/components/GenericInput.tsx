/**
 * TODO: this entire component needs to be refactored to use Attribute as a passed base
 * to then understand the type and value types of what attribute we're rendering with
 * what input and make the types all play nicely. At least now we have an idea of what
 * everything is!
 */

import * as React from 'react';

import {
  Checkbox,
  DatePicker,
  DateTimePicker,
  Icon,
  JSONInput,
  NumberInput,
  SingleSelect,
  SingleSelectOption,
  Textarea,
  TextInput,
  TimePicker,
  ToggleInput,
} from '@strapi/design-system';
import { Eye, EyeStriked } from '@strapi/icons';
import formatISO from 'date-fns/formatISO';
import isEqual from 'lodash/isEqual';
import { useIntl } from 'react-intl';

import { FieldSchema, useFieldHint } from '../hooks/useFieldHint';
import { useFocusInputField } from '../hooks/useFocusInputField';
import { pxToRem } from '../utils/pxToRem';

import type { TranslationMessage } from '../types';
import type { Attribute } from '@strapi/types';

interface InputOption {
  metadatas: {
    intlLabel: TranslationMessage;
    disabled: boolean;
    hidden: boolean;
  };
  key: string;
  value: string;
}

interface CustomInputProps<TAttribute extends Attribute.Any>
  extends Omit<GenericInputProps<TAttribute>, 'customInputs'> {
  ref?: React.Ref<HTMLElement>;
  hint?: string | React.JSX.Element | (string | React.JSX.Element)[];
}

export interface GenericInputProps<TAttribute extends Attribute.Any = Attribute.Any> {
  attribute?: TAttribute;
  autoComplete?: string;
  customInputs?: Record<string, React.ComponentType<CustomInputProps<TAttribute>>>;
  description?: TranslationMessage;
  disabled?: boolean;
  error?: string | TranslationMessage;
  intlLabel: TranslationMessage;
  labelAction?: React.ReactNode;
  name: string;
  onChange: (
    payload: {
      target: {
        name: string;
        value: Attribute.GetValue<TAttribute>;
        type?: string;
      };
    },
    shouldSetInitialValue?: boolean
  ) => void;
  options?: InputOption[];
  placeholder?: TranslationMessage;
  required?: boolean;
  step?: number;
  type: string;
  // TODO: The value depends on the input type, too complicated to handle all cases here
  value?: Attribute.GetValue<TAttribute>;
  isNullable?: boolean;
}

const GenericInput = ({
  autoComplete,
  customInputs,
  description,
  disabled,
  intlLabel,
  labelAction,
  error,
  name,
  onChange,
  options = [],
  placeholder,
  required,
  step,
  type,
  value: defaultValue,
  isNullable,
  attribute,
  ...rest
}: GenericInputProps) => {
  const { formatMessage } = useIntl();

  // TODO: Workaround to get the field hint values if they exist on the type
  const getFieldHintValue = (attribute?: Attribute.Any, key?: keyof FieldSchema) => {
    if (!attribute) return;

    if (key === 'minLength' && key in attribute) {
      return attribute[key];
    }

    if (key === 'maxLength' && key in attribute) {
      return attribute[key];
    }

    if (key === 'max' && key in attribute) {
      return attribute[key];
    }

    if (key === 'min' && key in attribute) {
      return attribute[key];
    }
  };

  const { hint } = useFieldHint({
    description,
    fieldSchema: {
      minLength: getFieldHintValue(attribute, 'minLength'),
      maxLength: getFieldHintValue(attribute, 'maxLength'),
      max: getFieldHintValue(attribute, 'max'),
      min: getFieldHintValue(attribute, 'min'),
    },
    type: attribute?.type || type,
  });

  const [showPassword, setShowPassword] = React.useState(false);
  const fieldRef = useFocusInputField(name);

  const CustomInput = customInputs ? customInputs[type] : null;

  // the API always returns null, which throws an error in React,
  // therefore we cast this case to undefined
  const value = defaultValue ?? undefined;

  /*
   TODO: ideally we should pass in `defaultValue` and `value` for
   inputs, in order to make them controlled components. This variable
   acts as a fallback for now, to prevent React errors in devopment mode

   See: https://github.com/strapi/strapi/pull/12861
  */
  const valueWithEmptyStringFallback = value ?? '';

  function getErrorMessage(error: string | TranslationMessage | undefined) {
    if (!error) {
      return null;
    }

    if (typeof error === 'string') {
      return formatMessage({ id: error, defaultMessage: error });
    }

    const values = {
      ...error.values,
    };

    return formatMessage(
      {
        id: error.id,
        defaultMessage: error?.defaultMessage ?? error.id,
      },
      values
    );
  }

  const errorMessage = getErrorMessage(error) ?? undefined;

  if (CustomInput) {
    return (
      <CustomInput
        {...rest}
        ref={fieldRef}
        attribute={attribute}
        description={description}
        hint={hint}
        disabled={disabled}
        intlLabel={intlLabel}
        labelAction={labelAction}
        error={errorMessage || ''}
        name={name}
        onChange={onChange}
        options={options}
        required={required}
        placeholder={placeholder}
        type={type}
        value={value}
      />
    );
  }

  const label = intlLabel.id
    ? formatMessage(
        { id: intlLabel.id, defaultMessage: intlLabel.defaultMessage },
        { ...intlLabel.values }
      )
    : name;

  const formattedPlaceholder = placeholder
    ? formatMessage(
        { id: placeholder.id, defaultMessage: placeholder.defaultMessage },
        { ...placeholder.values }
      )
    : '';

  switch (type) {
    case 'json': {
      return (
        <JSONInput
          // @ts-expect-error JSONInput ref is weird but it does work
          ref={fieldRef}
          label={label}
          labelAction={labelAction}
          value={value}
          error={errorMessage}
          disabled={disabled}
          hint={hint}
          required={required}
          onChange={(json) => {
            // Default to null when the field is not required and there is no input value
            const value =
              attribute && 'required' in attribute && !attribute?.required && !json.length
                ? null
                : json;
            onChange({ target: { name, value } }, false);
          }}
          minHeight={pxToRem(252)}
          maxHeight={pxToRem(504)}
        />
      );
    }
    case 'bool': {
      return (
        <ToggleInput
          ref={fieldRef}
          checked={defaultValue === null ? null : defaultValue || false}
          disabled={disabled}
          hint={hint}
          label={label}
          error={errorMessage}
          labelAction={labelAction}
          name={name}
          offLabel={formatMessage({
            id: 'app.components.ToggleCheckbox.off-label',
            defaultMessage: 'False',
          })}
          onLabel={formatMessage({
            id: 'app.components.ToggleCheckbox.on-label',
            defaultMessage: 'True',
          })}
          onChange={(e) => {
            onChange({ target: { name, value: e.target.checked } });
          }}
          required={required}
          onClear={() => {
            onChange({ target: { name, value: null } });
          }}
          clearLabel={
            isNullable
              ? formatMessage({
                  id: 'app.components.ToggleCheckbox.clear-label',
                  defaultMessage: 'Clear',
                })
              : undefined
          }
        />
      );
    }
    case 'checkbox': {
      return (
        <Checkbox
          ref={fieldRef}
          disabled={disabled}
          error={errorMessage}
          hint={hint}
          id={name}
          name={name}
          onValueChange={(value) => {
            onChange({ target: { name, value } });
          }}
          required={required}
          value={Boolean(value)}
        >
          {label}
        </Checkbox>
      );
    }
    case 'datetime': {
      return (
        <DateTimePicker
          ref={fieldRef}
          clearLabel={formatMessage({ id: 'clearLabel', defaultMessage: 'Clear' })}
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={(date) => {
            // check if date is not null or undefined
            const formattedDate = date ? date.toISOString() : null;

            onChange({ target: { name, value: formattedDate, type } });
          }}
          onClear={() => onChange({ target: { name, value: null, type } })}
          placeholder={formattedPlaceholder}
          required={required}
          value={value}
        />
      );
    }
    case 'date': {
      return (
        <DatePicker
          ref={fieldRef}
          clearLabel={formatMessage({ id: 'clearLabel', defaultMessage: 'Clear' })}
          disabled={disabled}
          error={errorMessage}
          label={label}
          id={name}
          hint={hint}
          name={name}
          onChange={(date) => {
            onChange({
              target: {
                name,
                value: date ? formatISO(date, { representation: 'date' }) : null,
                type,
              },
            });
          }}
          onClear={() => onChange({ target: { name, value: null, type } })}
          placeholder={formattedPlaceholder}
          required={required}
          selectedDate={value}
        />
      );
    }
    case 'number': {
      return (
        <NumberInput
          ref={fieldRef}
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onValueChange={(value) => {
            onChange({ target: { name, value, type } });
          }}
          placeholder={formattedPlaceholder}
          required={required}
          step={step}
          value={value}
        />
      );
    }
    case 'email': {
      return (
        <TextInput
          ref={fieldRef}
          autoComplete={autoComplete}
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={(e) => {
            onChange({ target: { name, value: e.target.value, type } });
          }}
          placeholder={formattedPlaceholder}
          required={required}
          type="email"
          value={valueWithEmptyStringFallback}
        />
      );
    }
    case 'timestamp':
    case 'text':
    case 'string': {
      return (
        <TextInput
          ref={fieldRef}
          autoComplete={autoComplete}
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={(e) => {
            onChange({ target: { name, value: e.target.value, type } });
          }}
          placeholder={formattedPlaceholder}
          required={required}
          type="text"
          value={valueWithEmptyStringFallback}
        />
      );
    }
    case 'password': {
      return (
        <TextInput
          ref={fieldRef}
          autoComplete={autoComplete}
          disabled={disabled}
          error={errorMessage}
          endAction={
            <button
              aria-label={formatMessage({
                id: 'Auth.form.password.show-password',
                defaultMessage: 'Show password',
              })}
              onClick={() => {
                setShowPassword((prev) => !prev);
              }}
              style={{
                border: 'none',
                padding: 0,
                background: 'transparent',
              }}
              type="button"
            >
              {showPassword ? (
                <Icon as={Eye} color="neutral500" />
              ) : (
                <Icon as={EyeStriked} color="neutral500" />
              )}
            </button>
          }
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={(e) => {
            onChange({ target: { name, value: e.target.value, type } });
          }}
          placeholder={formattedPlaceholder}
          required={required}
          type={showPassword ? 'text' : 'password'}
          value={valueWithEmptyStringFallback}
        />
      );
    }
    case 'select': {
      return (
        <SingleSelect
          ref={fieldRef}
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={(value) => {
            onChange({ target: { name, value, type: 'select' } });
          }}
          placeholder={formattedPlaceholder}
          required={required}
          value={value}
        >
          {options.map(({ metadatas: { intlLabel, disabled, hidden }, key, value }) => {
            return (
              <SingleSelectOption key={key} value={value} disabled={disabled} hidden={hidden}>
                {formatMessage(intlLabel)}
              </SingleSelectOption>
            );
          })}
        </SingleSelect>
      );
    }
    case 'textarea': {
      return (
        <Textarea
          ref={fieldRef}
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={(event) => onChange({ target: { name, value: event.target.value, type } })}
          required={required}
          placeholder={formattedPlaceholder}
          value={valueWithEmptyStringFallback}
        />
      );
    }
    case 'time': {
      let time = value;

      // The backend send a value which has the following format: '00:45:00.000'
      // or the time picker only supports hours & minutes so we need to mutate the value
      if (typeof value === 'string' && value.split(':').length > 2) {
        const [hour, minute] = value.split(':');
        time = `${hour}:${minute}`;
      }

      return (
        <TimePicker
          ref={fieldRef}
          clearLabel={formatMessage({ id: 'clearLabel', defaultMessage: 'Clear' })}
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={(time) => {
            onChange({ target: { name, value: `${time}`, type } });
          }}
          onClear={() => {
            onChange({ target: { name, value: null, type } });
          }}
          required={required}
          value={time}
        />
      );
    }
    default: {
      /**
       * If there's no component for the given type, we return a disabled text input
       * showing a "Not supported" title to illustrate the issue.
       */
      return (
        <TextInput
          disabled
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          placeholder="Not supported"
          required={required}
          type="text"
          value=""
        />
      );
    }
  }
};

/**
 * we've memoized this component because we use a context to store all the data in our form in the content-manager.
 * This then causes _every_ component to re-render because there are no selects incurring performance issues
 * in content-types as the content-type gets more complicated.
 */
const MemoizedGenericInput = React.memo(GenericInput, isEqual);

export { MemoizedGenericInput as GenericInput };
