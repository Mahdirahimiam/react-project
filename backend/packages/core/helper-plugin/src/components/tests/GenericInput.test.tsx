import { DesignSystemProvider } from '@strapi/design-system';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { IntlProvider } from 'react-intl';
import { MemoryRouter, MemoryRouterProps } from 'react-router-dom';

import { GenericInput, GenericInputProps } from '../GenericInput';

function renderField(
  props?: Partial<GenericInputProps>,
  { initialEntries }: Pick<MemoryRouterProps, 'initialEntries'> = {}
) {
  return {
    ...render(
      // @ts-expect-error - TODO: fix the Attribute issue.
      <GenericInput
        intlLabel={{
          id: 'label.test',
          defaultMessage: 'Default label',
        }}
        onChange={jest.fn}
        value={null}
        {...props}
      />,
      {
        wrapper: ({ children }) => (
          <IntlProvider locale="en" messages={{}}>
            <DesignSystemProvider locale="en-GB">
              <MemoryRouter initialEntries={initialEntries}>{children}</MemoryRouter>
            </DesignSystemProvider>
          </IntlProvider>
        ),
      }
    ),
    user: userEvent.setup(),
  };
}

describe('GenericInput', () => {
  describe('number', () => {
    const renderNumber = (props: Partial<GenericInputProps>) => {
      return renderField({
        type: 'number',
        name: 'number',
        placeholder: {
          id: 'placeholder.test',
          defaultMessage: 'Default placeholder',
        },
        required: true,
        ...props,
      });
    };

    test('renders an error message', () => {
      const { getByText } = renderNumber({ error: 'Error message' });
      expect(getByText('Error message')).toBeInTheDocument();
    });

    test('renders a number (int) value', () => {
      const { getByRole } = renderNumber({ value: 1 });
      expect(getByRole('textbox')).toHaveValue('1');
    });

    test('renders a number (float) value', () => {
      const { getByRole } = renderNumber({ value: 1.3333 });
      expect(getByRole('textbox')).toHaveValue('1.3333');
    });

    test('does not call onChange callback on first render', () => {
      const spy = jest.fn();
      renderNumber({ value: null, onChange: spy });
      expect(spy).not.toHaveBeenCalled();
    });

    test('does not call onChange callback if the value does not change', async () => {
      const spy = jest.fn();
      const { getByRole, user } = renderNumber({ value: 23, onChange: spy });

      await user.type(getByRole('textbox'), '23');

      expect(spy).not.toHaveBeenCalledWith();
    });

    test('does call onChange callback with number (int) value', async () => {
      const spy = jest.fn();
      const { user, getByRole } = renderNumber({ value: null, onChange: spy });

      await user.type(getByRole('textbox'), '23');

      expect(spy).toHaveBeenCalledWith({ target: { name: 'number', type: 'number', value: 23 } });
    });

    test('does call onChange callback with number (float) value', async () => {
      const spy = jest.fn();
      const { getByRole, user } = renderNumber({ value: null, onChange: spy });

      await user.type(getByRole('textbox'), '1.3333');

      expect(spy).toHaveBeenCalledWith({
        target: { name: 'number', type: 'number', value: 1.3333 },
      });
    });

    test('does call onChange callback with number (0) value', async () => {
      const spy = jest.fn();
      const { getByRole, user } = renderNumber({ value: null, onChange: spy });

      await user.type(getByRole('textbox'), '0');

      expect(spy).toHaveBeenCalledWith({
        target: { name: 'number', type: 'number', value: 0 },
      });
    });
  });

  describe('json', () => {
    test('renders and matches the snapshot', () => {
      const { container } = renderField({
        type: 'json',
        name: 'json',
        value: null,
      });

      expect(container).toMatchSnapshot();
    });
  });

  describe('date', () => {
    const renderDate = (props?: Partial<GenericInputProps>) => {
      return renderField({
        type: 'date',
        name: 'date',
        intlLabel: {
          id: 'label.test',
          defaultMessage: 'date',
        },
        ...props,
      });
    };

    it('should allow the user to clear the field', async () => {
      const onChange = jest.fn();

      const { getByRole, user } = renderDate({
        value: new Date(),
        onChange,
      });

      await user.click(getByRole('button', { name: 'Clear' }));

      expect(getByRole('combobox', { name: 'date' })).toHaveValue('');
      expect(onChange).toHaveBeenCalledWith({
        target: { name: 'date', type: 'date', value: null },
      });
    });
  });

  describe('datetime', () => {
    const renderDateTime = (props?: Partial<GenericInputProps>) => {
      return renderField({
        type: 'datetime',
        name: 'datetime-picker',
        intlLabel: {
          id: 'label.test',
          defaultMessage: 'datetime picker',
        },
        ...props,
      });
    };

    test('renders the datetime picker with the correct value for date and time', async () => {
      const { getByRole, user } = renderDateTime();

      await user.click(getByRole('combobox', { name: 'Choose date' }));
      await user.click(getByRole('gridcell', { name: /15/ }));

      const today = new Date().setDate(15);
      const formattedDate = new Intl.DateTimeFormat('en-GB').format(today);
      expect(getByRole('combobox', { name: 'Choose date' })).toHaveValue(formattedDate);
      expect(getByRole('combobox', { name: 'Choose time' })).toHaveValue('00:00');
    });

    test('simulate clicking on the Clear button in the date and check if the date and time are empty', async () => {
      const { getByRole, user } = renderDateTime();

      await user.click(getByRole('combobox', { name: 'Choose date' }));
      await user.click(getByRole('gridcell', { name: /15/ }));
      await user.click(getByRole('button', { name: 'Clear date' }));

      expect(getByRole('combobox', { name: 'Choose date' })).toHaveValue('');
      expect(getByRole('combobox', { name: 'Choose time' })).toHaveValue('');
    });
  });

  describe('textarea', () => {
    test('calls onChange callback with textarea value', async () => {
      const onChange = jest.fn();

      const { getByRole, user } = renderField({
        type: 'textarea',
        name: 'textarea',
        onChange,
      });

      const textarea = getByRole('textbox');

      await user.type(textarea, 'test');

      expect(onChange).toHaveBeenNthCalledWith(1, {
        target: { name: 'textarea', type: 'textarea', value: 't' },
      });
      expect(onChange).toHaveBeenNthCalledWith(2, {
        target: { name: 'textarea', type: 'textarea', value: 'e' },
      });
      expect(onChange).toHaveBeenNthCalledWith(3, {
        target: { name: 'textarea', type: 'textarea', value: 's' },
      });
      expect(onChange).toHaveBeenNthCalledWith(4, {
        target: { name: 'textarea', type: 'textarea', value: 't' },
      });
    });
  });

  describe('auto-focussing', () => {
    test.each([
      'bool',
      'checkbox',
      'date',
      'datetime',
      'email',
      'number',
      'password',
      'select',
      'text',
      'textarea',
      'time',
    ] as const)('auto-focuses on %s', (type) => {
      const { getByLabelText } = renderField(
        { type, name: 'test' },
        { initialEntries: [{ pathname: '/', search: `field=test` }] }
      );

      /**
       * datetime renders two fields, and it should focus the date field.
       */

      if (type === 'datetime') {
        expect(getByLabelText('Choose date')).toHaveFocus();
      } else {
        expect(getByLabelText('Default label')).toHaveFocus();
      }
    });
  });
});
