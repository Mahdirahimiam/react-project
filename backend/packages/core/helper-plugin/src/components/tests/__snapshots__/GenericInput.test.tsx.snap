// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericInput json renders and matches the snapshot 1`] = `
.c6 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c1 {
  font-size: 0.75rem;
  line-height: 1.33;
  font-weight: 600;
  color: #32324d;
}

.c3 {
  font-size: 0.875rem;
  border-radius: 4px;
  max-height: 31.5rem;
  min-height: 15.75rem;
}

.c0 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 4px;
}

.c4 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  line-height: 1.25;
}

.c5 .cm-editor {
  background-color: #32324d;
  width: 100%;
  outline: none;
}

.c5 .cm-scroller {
  border: 1px solid #dcdce4;
  outline: none;
  box-shadow: 0;
  -webkit-transition-property: border-color,box-shadow,fill;
  transition-property: border-color,box-shadow,fill;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
}

.c5 .cm-scroller:focus-within {
  border: 1px solid #4945ff;
  box-shadow: #4945ff 0px 0px 0px 2px;
}

.c5 .cm-editor,
.c5 .cm-scroller {
  border-radius: 4px;
}

.c5 .cm-gutters,
.c5 .cm-activeLineGutter {
  background-color: #4a4a6a;
}

<div>
  <div
    class=""
  >
    <div
      class="c0"
    >
      <label
        class="c1 c2"
        for=":rg:"
      >
        Default label
      </label>
      <div
        class="c3 c4 c5"
        font-size="2"
      >
        <div
          class="cm-editor ͼ1 ͼ3 ͼ4 ͼ19 ͼo"
        >
          <div
            aria-live="polite"
            style="position: fixed; top: -10000px;"
          />
          <div
            class="cm-scroller"
            tabindex="-1"
          >
            <div
              aria-hidden="true"
              class="cm-gutters"
              style="min-height: 14px; position: sticky;"
            >
              <div
                class="cm-gutter cm-lineNumbers"
              >
                <div
                  class="cm-gutterElement"
                  style="height: 0px; visibility: hidden; pointer-events: none;"
                >
                  9
                </div>
                <div
                  class="cm-gutterElement cm-activeLineGutter"
                  style="height: 14px;"
                >
                  1
                </div>
              </div>
              <div
                class="cm-gutter cm-foldGutter"
              >
                <div
                  class="cm-gutterElement"
                  style="height: 0px; visibility: hidden; pointer-events: none;"
                >
                  <span
                    title="Unfold line"
                  >
                    ›
                  </span>
                </div>
                <div
                  class="cm-gutterElement cm-activeLineGutter"
                  style="height: 14px;"
                />
              </div>
            </div>
            <div
              aria-autocomplete="list"
              aria-multiline="true"
              autocapitalize="off"
              autocorrect="off"
              class="cm-content"
              contenteditable="true"
              data-language="json"
              role="textbox"
              spellcheck="false"
              style="tab-size: 4"
              translate="no"
            >
              <div
                class="cm-activeLine cm-line"
              >
                <br />
              </div>
            </div>
            <div
              aria-hidden="true"
              class="cm-layer cm-layer-above cm-cursorLayer"
              style="z-index: 150; animation-duration: 1200ms;"
            />
            <div
              aria-hidden="true"
              class="cm-layer cm-selectionLayer"
              style="z-index: -2;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="c6"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
